halo2_gadgets/src/sinsemilla/chip/hash_to_point.rs

```diff
@@ -68,29 +66,85 @@ where
         ))
     }
 
+    /// [Specification](https://p.z.cash/halo2-0.1:sinsemilla-constraints?partial).
+    #[allow(non_snake_case)]
+    #[allow(clippy::type_complexity)]
+    pub(super) fn hash_message_with_private_init(
+        &self,
+        region: &mut Region<'_, pallas::Base>,
+        Q: &NonIdentityEccPoint,
+        message: &<Self as SinsemillaInstructions<
+            pallas::Affine,
+            { sinsemilla::K },
+            { sinsemilla::C },
+        >>::Message,
+    ) -> Result<
+        (
+            NonIdentityEccPoint,
+            Vec<Vec<AssignedCell<pallas::Base, pallas::Base>>>,
+        ),
+        Error,
+    > {
+        if !self.config().init_from_private_point {
+            return Err(Error::IllegalHashFromPrivatePoint);
+        }
+
+        let (offset, x_a, y_a) = self.private_q_initialization(region, Q)?;
+
+        let (x_a, y_a, zs_sum) = self.hash_all_pieces(region, offset, message, x_a, y_a)?;
+
+        #[cfg(test)]
+        self.check_hash_result(EccPointQ::PrivatePoint(Q.clone()), message, &x_a, &y_a);
+
+        x_a.value()
+            .zip(y_a.value())
+            .error_if_known_and(|(x_a, y_a)| x_a.is_zero_vartime() || y_a.is_zero_vartime())?;
+        Ok((
+            NonIdentityEccPoint::from_coordinates_unchecked(x_a.0, y_a),
+            zs_sum,
+        ))
+    }
+
     #[allow(non_snake_case)]
-    /// Assign the coordinates of the initial public point `Q`
-    /// - `x_Q` in a advice column, and
-    /// - `y_Q` in a fixed column.
+    /// Assign the coordinates of the initial public point `Q`.
     ///
+    /// If init_from_private_point is not set,
     /// | offset | x_A | q_sinsemilla4 | fixed_y_q |
     /// --------------------------------------
     /// |   0    | x_Q |   1           |   y_Q     |
+    ///
+    /// If init_from_private_point is set,
+    /// | offset | x_A | x_P | q_sinsemilla4 |
+    /// --------------------------------------
+    /// |   0    |     | y_Q |               |
+    /// |   1    | x_Q |     |         1     |
```

Comment:
This should be documented in `book/src/design/gadgets/sinsemilla.md`. I have a commit for that and will push it tomorrow.